### ISSUE TITLE
Hyperopt always sampling the same points with TPE strategy

### ISSUE THREAD

**Issue by mikolchon** [hyperopt/hyperopt#276]
I want to do a convergence study of Ridge regression with increasing training size. For this I train the Ridge regressor using `hyperopt` for different training sizes `[10, 20, 40, 80]`. As far as I understand, the TPE strategy adapts to the optimization, so I would expect that the sampling for each training with different training sizes be different, but that's not the case. Here is a code that shows that TPE is always sampling the same points:

    import numpy as np
    X = np.random.randn(100,1)
    Y = np.random.randn(100,1)

    from sklearn.model_selection import train_test_split
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y,
                                                        test_size=0.2,
                                                        random_state=0)

    from sklearn.linear_model import Ridge
    from hyperopt import fmin, tpe, hp, Trials

    results = []
    for n in [10, 20, 40, 80]:
        print "\nTraining with", n, "samples"

        def objective(alpha):
            model = Ridge(alpha=alpha).fit(X_train[:n], Y_train[:n])
            return -model.score(X_test, Y_test)  # negative so that lower means better

        trials = Trials()
        best = fmin(objective, hp.uniform('alpha', 0.0, 10.0),
                    algo=tpe.suggest,
                    max_evals=5,
                    trials=trials)

        print "alphas sampled:"
        for trial in trials:
            print trial['misc']['vals']

The output of this is

```
Training with 10 samples
alphas tried:
{'alpha': [6.964691855978616]}
{'alpha': [1.0606490595473272]}
{'alpha': [5.067260114995699]}
{'alpha': [1.0517212721767522]}
{'alpha': [5.238333199546759]}

Training with 20 samples
alphas tried:
{'alpha': [6.964691855978616]}
{'alpha': [1.0606490595473272]}
{'alpha': [5.067260114995699]}
{'alpha': [1.0517212721767522]}
{'alpha': [5.238333199546759]}

Training with 40 samples
alphas tried:
{'alpha': [6.964691855978616]}
{'alpha': [1.0606490595473272]}
{'alpha': [5.067260114995699]}
{'alpha': [1.0517212721767522]}
{'alpha': [5.238333199546759]}

Training with 80 samples
alphas tried:
{'alpha': [6.964691855978616]}
{'alpha': [1.0606490595473272]}
{'alpha': [5.067260114995699]}
{'alpha': [1.0517212721767522]}
{'alpha': [5.238333199546759]}

```

**Comment by robertmuil** [comment]
Also very interested in an answer to this!

**Comment by jaberg** [comment]
The TPE algorithm is meant to start off with a few random points to
calibrate what threshold to use for the first non-random iterations. I
think all SMBO algorithms work this way. Five evaluations is under the
initial number of random trials. If you run with a greater value of
max_evals, and the functions you're comparing are different-enough, then
you should see the search sequences diverge.

**Comment by robertmuil** [comment]
Makes complete sense! Thanks very much for speedy response... and for the package itself, James, superb work.

**Comment by brendonboshell** [comment]
@mikolchon I run the script you provided, but it did not reproduce this problem. (In your code you have "alphas sampled" but the output is "alphas tried" so I suspect you have posted the wrong script?)

**Comment by mikolchon** [comment]
No, actually I changed that later but the rest remains the same.

**Comment by brendonboshell** [comment]
@mikolchon Ah okay. I guess the package has been updated or there is something different in my environment. But I get the following output when I run your script, for example:

```
Training with 10 samples
alphas sampled:
{'alpha': [3.79133498788343]}
{'alpha': [3.362268365630138]}
{'alpha': [7.932068239025699]}
{'alpha': [5.747214681543328]}
{'alpha': [4.885889830800777]}

Training with 20 samples
alphas sampled:
{'alpha': [8.126363133987985]}
{'alpha': [3.6928080135823005]}
{'alpha': [6.0615003266282095]}
{'alpha': [0.12258599601074671]}
{'alpha': [3.3106451834390818]}

Training with 40 samples
alphas sampled:
{'alpha': [7.379141631699014]}
{'alpha': [1.1828223991675668]}
{'alpha': [1.9082652620125984]}
{'alpha': [8.28440316150725]}
{'alpha': [8.273823772355344]}

Training with 80 samples
alphas sampled:
{'alpha': [7.582625806488121]}
{'alpha': [5.073513930133933]}
{'alpha': [0.44799594354303474]}
{'alpha': [1.7551078540142928]}
{'alpha': [8.946430951909248]}
```

**Comment by vertti** [comment]
I'm also encountering this, it seems the first iteration always has the same parameter values..

**Comment by abiteboul** [comment]
@vertti This is the expected behaviour for the TPE algorithm, see the above explanation by @jaberg:

> The TPE algorithm is meant to start off with a few random points to calibrate what threshold to use for the first non-random iterations. I think all SMBO algorithms work this way. Five evaluations is under the
initial number of random trials. If you run with a greater value of max_evals, and the functions you're comparing are different-enough, then you should see the search sequences diverge.

This issue can be closed.
